File: cli/internal/context/hook.go
package context

import (
	"bufio"
	"bytes"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"

	"harness-data/cli/internal/harness"
	"harness-data/cli/internal/sessionstate"
)

type Output struct {
	HookSpecificOutput HookSpecificOutput `json:"hookSpecificOutput"`
}

type HookSpecificOutput struct {
	HookEventName     string `json:"hookEventName"`
	AdditionalContext string `json:"additionalContext"`
}

type promptPayload struct {
	SessionID string `json:"session_id"`
	Prompt    string `json:"prompt"`
}

type timeContext struct {
	CurrentDate string `json:"current_date"`
	Timezone    string `json:"timezone"`
	TimePolicy  string `json:"time_policy"`
	Prompt      string `json:"prompt"`
}

func RunClaudeHook(root string, input []byte) (bool, Output, error) {
	payload, ok := parsePromptPayload(input)
	if !ok || payload.Prompt == "" {
		return false, Output{}, nil
	}
	prompt := payload.Prompt
	resolver, err := harness.NewPathResolver(root)
	if err != nil {
		return false, Output{}, err
	}
	tc := buildTimeContext(prompt, resolver)
	response, plan, err := BuildWithPlan(root, prompt)
	if err != nil {
		return false, Output{}, err
	}
	additionalContext, err := buildWikiAdditionalContext(tc, response, plan)
	if err != nil {
		return false, Output{}, err
	}
	sessionID := hookSessionID(payload)
	if err := writeWikiPlanState(root, sessionID, prompt, plan); err != nil {
		return false, Output{}, err
	}
	if os.Getenv("QDM_HARNESS_DIAG") == "1" {
		if err := recordDiagnostic(root, sessionID, prompt, additionalContext, tc, response); err != nil {
			return false, Output{}, err
		}
	}
	return true, Output{
		HookSpecificOutput: HookSpecificOutput{
			HookEventName:     "UserPromptSubmit",
			AdditionalContext: additionalContext,
		},
	}, nil
}

func buildWikiAdditionalContext(tc timeContext, response harness.ContextResponse, plan WikiPlan) (string, error) {
	timeJSON, err := json.Marshal(tc)
	if err != nil {
		return "", err
	}
	var b strings.Builder
	b.WriteString("# Data Harness Context\n\n")
	b.WriteString("时间解析 JSON：`")
	b.Write(timeJSON)
	b.WriteString("`\n\n")
	b.WriteString("Harness mode: ")
	b.WriteString(plan.Mode)
	b.WriteString("\n")
	if plan.SelectedPlaybook != "" {
		b.WriteString("selectedPlaybook: ")
		b.WriteString(plan.SelectedPlaybook)
		b.WriteString("\n")
	}
	if plan.SelectedTemplate != "" {
		b.WriteString("selectedTemplate: ")
		b.WriteString(plan.SelectedTemplate)
		b.WriteString("\n")
	}
	if plan.TemplateSelection.Status != "" {
		b.WriteString("templateSelection: ")
		b.WriteString(plan.TemplateSelection.Status)
		if plan.TemplateSelection.Reason != "" {
			b.WriteString(" (")
			b.WriteString(plan.TemplateSelection.Reason)
			b.WriteString(")")
		}
		b.WriteString("\n")
	}
	if len(plan.SelectedPlaybooks) > 0 {
		b.WriteString("selectedPlaybooks:\n")
		for _, playbook := range plan.SelectedPlaybooks {
			b.WriteString("- ")
			b.WriteString(playbook.Path)
			b.WriteString("\n")
		}
	}
	if plan.Mode == sessionstate.ModeFree {
		b.WriteString("reason: ")
		b.WriteString(plan.Reason)
		b.WriteString("\n")
	}
	b.WriteString("\n必须先读取以下 contextFiles：\n")
	for _, ref := range response.ContextFiles {
		b.WriteString("- ")
		b.WriteString(ref.Path)
		if ref.Reason != "" {
			b.WriteString(" (")
			b.WriteString(ref.Reason)
			b.WriteString(")")
		}
		b.WriteString("\n")
	}
	b.WriteString("\nInstruction: ")
	b.WriteString(response.Instruction)
	b.WriteString("\n\nConstraints:\n")
	for _, constraint := range response.Constraints {
		b.WriteString("- ")
		b.WriteString(constraint)
		b.WriteString("\n")
	}
	return b.String(), nil
}

func parsePromptPayload(input []byte) (promptPayload, bool) {
	var payload promptPayload
	if err := json.Unmarshal(input, &payload); err != nil {
		return promptPayload{}, false
	}
	return payload, true
}

func buildTimeContext(prompt string, resolver harness.PathResolver) timeContext {
	currentDate := os.Getenv("QDM_HARNESS_CURRENT_DATE")
	current, err := time.Parse("2006-01-02", currentDate)
	if err != nil {
		current = time.Now()
	}
	timezone := os.Getenv("QDM_HARNESS_TIMEZONE")
	if timezone == "" {
		timezone = os.Getenv("TZ")
	}
	if timezone == "" {
		timezone = "Asia/Shanghai"
	}
	timePolicy := "spec/common/time-policy.md"
	if info, err := os.Stat(resolver.Resolve("rules/QDM 时间口径/spec.md")); err == nil && !info.IsDir() {
		timePolicy = "rules/QDM 时间口径/spec.md"
	}
	return timeContext{
		CurrentDate: current.Format("2006-01-02"),
		Timezone:    timezone,
		TimePolicy:  fmt.Sprintf("Use %s to infer --date, --week, or --month. Do not use date ranges.", resolver.ResolveRel(timePolicy)),
		Prompt:      prompt,
	}
}

func recordDiagnostic(root, sessionID, prompt, context string, tc timeContext, response harness.ContextResponse) error {
	event := map[string]any{
		"ts":              time.Now().UTC().Format(time.RFC3339Nano),
		"session_id":      sessionID,
		"event":           "user_prompt_context",
		"matched_domains": diagnosticMatchedDomains(response.ContextFiles),
		"context_files":   contextFileDiagnostics(root, response.ContextFiles),
		"keyword_hits":    keywordHits(response.ContextFiles),
		"prompt_bytes":    len([]byte(prompt)),
		"context_bytes":   len([]byte(context)),
		"context_lines":   countLines(context),
		"time_context":    tc,
	}
	data, err := json.Marshal(event)
	if err != nil {
		return err
	}
	dir := sessionstate.DiagnosticsDir(root)
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return err
	}
	path := filepath.Join(dir, sessionstate.SafeSessionID(sessionID)+".jsonl")
	file, err := os.OpenFile(path, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0o644)
	if err != nil {
		return err
	}
	defer file.Close()
	_, err = file.Write(append(data, '\n'))
	return err
}

func contextFileDiagnostics(root string, refs []harness.FileRef) []map[string]any {
	resolver, err := harness.NewPathResolver(root)
	if err != nil {
		resolver = harness.PathResolver{Root: root, Paths: harness.PathsConfig{Spec: "spec", Routing: "routing", Playbooks: "playbooks", Templates: "templates"}}
	}
	var out []map[string]any
	for _, ref := range refs {
		rel := ref.Path
		path := resolver.Resolve(rel)
		stats := pathStats(path)
		stats["relative_path"] = rel
		stats["reason"] = ref.Reason
		out = append(out, stats)
	}
	return out
}

func pathStats(path string) map[string]any {
	data, err := os.ReadFile(path)
	if err != nil {
		return map[string]any{"path": path, "exists": false, "bytes": 0, "lines": 0}
	}
	return map[string]any{
		"path":   path,
		"exists": true,
		"bytes":  len(data),
		"lines":  countLinesBytes(data),
	}
}

func countLines(s string) int {
	if s == "" {
		return 0
	}
	return countLinesBytes([]byte(s))
}

func countLinesBytes(data []byte) int {
	if len(data) == 0 {
		return 0
	}
	lines := bytes.Count(data, []byte{'\n'})
	if data[len(data)-1] != '\n' {
		lines++
	}
	return lines
}

func diagnosticMatchedDomains(refs []harness.FileRef) []string {
	seen := map[string]bool{}
	for _, ref := range refs {
		parts := strings.Split(ref.Path, "/")
		if len(parts) < 3 || parts[1] == "common" {
			continue
		}
		seen[parts[1]] = true
	}
	var domains []string
	for domain := range seen {
		domains = append(domains, domain)
	}
	sort.Strings(domains)
	return domains
}

func hookSessionID(payload promptPayload) string {
	if payload.SessionID != "" {
		return payload.SessionID
	}
	sessionID := os.Getenv("CLAUDE_SESSION_ID")
	if sessionID == "" {
		return "unknown"
	}
	return sessionID
}

func writeWikiPlanState(root, sessionID, prompt string, plan WikiPlan) error {
	state, err := sessionstate.Load(root, sessionID)
	if err != nil {
		return err
	}
	state.Mode = plan.Mode
	state.Prompt = prompt
	state.StartedAt = time.Now().UTC().Format(time.RFC3339Nano)
	state.PlaybookCandidates = plan.Candidates
	state.SelectedPlaybook = ""
	state.SelectedTemplate = ""
	state.SelectedPlaybooks = nil
	state.Composite = nil
	state.Reason = ""
	state.TemplateInjected = false
	state.Reports = map[string]*sessionstate.Report{}
	switch plan.Mode {
	case sessionstate.ModeSingle:
		state.SelectedPlaybook = plan.SelectedPlaybook
		state.SelectedTemplate = plan.SelectedTemplate
	case sessionstate.ModeMulti:
		state.SelectedPlaybooks = append([]sessionstate.PlaybookCandidate{}, plan.SelectedPlaybooks...)
	case sessionstate.ModeReport:
		state.SelectedPlaybook = plan.SelectedPlaybook
		state.SelectedTemplate = plan.SelectedTemplate
	case sessionstate.ModeFree:
		state.Reason = plan.Reason
	}
	return sessionstate.Save(root, sessionID, state)
}

func keywordHits(refs []harness.FileRef) []map[string]string {
	var hits []map[string]string
	for _, ref := range refs {
		if !strings.Contains(ref.Reason, "keyword: ") {
			continue
		}
		parts := strings.SplitN(ref.Reason, "keyword: ", 2)
		if len(parts) != 2 || parts[1] == "" {
			continue
		}
		hits = append(hits, map[string]string{"path": ref.Path, "keyword": parts[1]})
	}
	return hits
}

func ReadHookStdin() ([]byte, error) {
	var buf bytes.Buffer
	scanner := bufio.NewScanner(os.Stdin)
	for scanner.Scan() {
		buf.Write(scanner.Bytes())
		buf.WriteByte('\n')
	}
	return buf.Bytes(), scanner.Err()
}

File: packages/data-harness-cli/src/lib/context/hook.js
import { createHash } from "node:crypto";
import { appendFileSync, mkdirSync, statSync } from "node:fs";
import path from "node:path";

import { loadConfig, newPathResolver } from "../harness.js";
import { contextFromHookPayload, RootContextError } from "../root-context.js";
import { assertWorkspaceAllowed } from "../workspace-policy.js";
import { diagnosticsDir, load as loadState, MODE_FREE, MODE_MULTI, MODE_REPORT, MODE_SINGLE, safeSessionId, save as saveState } from "../sessionstate.js";
import { buildWithPlan } from "./build.js";

const WORKBUDDY_SESSION_PREFIX = "workbuddy:";

export function runClaudeHook(root, input, context = null, options = {}) {
  const payload = parsePromptPayload(input);
  if (!payload || !payload.prompt) return { ok: false, output: null };
  let effectiveContext = context;
  try {
    effectiveContext = contextFromHookPayload(payload, {
      root,
      env: options.env || process.env,
      baseContext: context,
    }) || effectiveContext;
  } catch (error) {
    if (error instanceof RootContextError) {
      return {
        ok: true,
        output: workspaceRequiredOutput(`${error.code}: ${error.message}`),
      };
    }
    return {
      ok: true,
      output: hookSafetyOutput(
        "QDM_HARNESS_UNAVAILABLE: Harness hook context could not be loaded. Do not run qdm-metric-cli or estimate values until the runtime configuration is repaired.",
      ),
    };
  }
  const sessionID = hookSessionID(payload);
  const workspaceDecision = checkWorkspacePolicy(effectiveContext);
  if (workspaceDecision === "denied") return { ok: false, output: null };
  if (workspaceDecision === "unavailable") {
    return {
      ok: true,
      output: hookSafetyOutput("QDM_SETUP_REQUIRED: configure the Harness Data workspace allowlist before using this project."),
    };
  }
  if (
    isOnDemandContext(effectiveContext, payload.prompt, sessionID, root) &&
    !shouldAutoRecall(effectiveContext, payload.prompt)
  ) {
    return { ok: false, output: null };
  }
  if (effectiveContext && (!effectiveContext.workspaceRoot || effectiveContext.capabilities?.canWriteWorkspace === false)) {
    // Only report operations require a writable workspace; read-only metric
    // context stays available for read-only hosts (QwenPaw ships with
    // canWriteWorkspace=false but still needs the injected manuals).
    let requiresWorkspace = true;
    try {
      requiresWorkspace = buildWithPlan(effectiveContext, payload.prompt).mode === MODE_REPORT;
    } catch {
      requiresWorkspace = true; // fail closed when the plan cannot be built
    }
    if (requiresWorkspace) {
      return {
        ok: true,
        output: workspaceRequiredOutput("QDM_WORKSPACE_REQUIRED: explicit context/report operations require workspaceRoot; read-only context is still available."),
      };
    }
  }
  try {
    const persistState = !effectiveContext ||
      String(effectiveContext.host || "").trim().toLowerCase() !== "codex" ||
      isExplicitContextPrompt(payload.prompt) ||
      isContinuationPrompt(payload.prompt);
    return runPromptHook(root, payload.prompt, sessionID, effectiveContext, { persistState });
  } catch (error) {
    return {
      ok: true,
      output: hookSafetyOutput(
        `QDM_HARNESS_UNAVAILABLE: Harness context could not be built${error?.code ? ` (${error.code})` : ""}. Do not run qdm-metric-cli or estimate values until the runtime configuration is repaired.`,
      ),
    };
  }
}

export function runWorkBuddyHook(root, input, context = null, options = {}) {
  const payload = parsePromptPayload(input);
  if (!payload || !String(payload.prompt || "").trim()) return { ok: false, output: null };
  let effectiveContext = context;
  try {
    effectiveContext = contextFromHookPayload(payload, {
      root,
      env: options.env || process.env,
      baseContext: context,
    }) || effectiveContext;
  } catch (error) {
    if (error instanceof RootContextError) {
      return {
        ok: true,
        output: workBuddySafetyOutput(`${error.code}: ${error.message}`),
      };
    }
    throw error;
  }
  const workspaceDecision = checkWorkspacePolicy(effectiveContext);
  if (workspaceDecision === "denied") return { ok: false, output: null };
  if (workspaceDecision === "unavailable") {
    return {
      ok: true,
      output: workBuddySafetyOutput("QDM_SETUP_REQUIRED: configure the Harness Data workspace allowlist before using this project."),
    };
  }
  const sessionID = String(payload.session_id || "").trim();
  if (!sessionID) {
    return {
      ok: true,
      output: workBuddySafetyOutput(
        "QDM_HARNESS_BLOCKED: WorkBuddy did not provide a stable session_id. Do not run qdm-metric-cli, do not estimate data, and do not run template commands in this turn. Start a new WorkBuddy session or update WorkBuddy before retrying.",
      ),
    };
  }
  if (isOnDemandContext(effectiveContext, payload.prompt, WORKBUDDY_SESSION_PREFIX + sessionID, root)) {
    return { ok: false, output: null };
  }
  if (effectiveContext && (!effectiveContext.workspaceRoot || effectiveContext.capabilities?.canWriteWorkspace === false)) {
    return {
      ok: true,
      output: workspaceRequiredOutput("QDM_WORKSPACE_REQUIRED: explicit context/report operations require workspaceRoot; read-only context is still available."),
    };
  }
  let cfg;
  try {
    cfg = loadConfig(effectiveContext || root);
  } catch {
    return {
      ok: true,
      output: workBuddySafetyOutput(
        "QDM_HARNESS_UNAVAILABLE: Harness configuration could not be loaded. Do not run qdm-metric-cli or estimate data until the runtime configuration is repaired.",
      ),
    };
  }
  let result;
  try {
    result = runPromptHook(root, payload.prompt, WORKBUDDY_SESSION_PREFIX + sessionID, effectiveContext);
  } catch {
    return {
      ok: true,
      output: workBuddySafetyOutput(
        "QDM_HARNESS_UNAVAILABLE: Harness context could not be built. Do not run qdm-metric-cli, estimate values, or guess playbooks/templates in this turn.",
      ),
    };
  }
  if (!result.ok) return { ok: false, output: null };
  result.output.hookSpecificOutput.additionalContext =
    `authzMode: ${cfg.authz.mode}\n\n${result.output.hookSpecificOutput.additionalContext}`;
  return {
    ok: true,
    output: {
      continue: true,
      hookSpecificOutput: result.output.hookSpecificOutput,
    },
  };
}

function workBuddySafetyOutput(message) {
  return {
    continue: true,
    systemMessage: message,
    hookSpecificOutput: {
      hookEventName: "UserPromptSubmit",
      additionalContext: message,
    },
  };
}

function checkWorkspacePolicy(context) {
  if (!context?.workspaceRoot) return "ok";
  try {
    assertWorkspaceAllowed(context);
    return "ok";
  } catch (error) {
    if (error?.code === "QDM_WORKSPACE_NOT_ALLOWED") return "denied";
    if (error?.code === "QDM_SETUP_REQUIRED") return "unavailable";
    throw error;
  }
}

function workspaceRequiredOutput(message) {
  return hookSafetyOutput(message);
}

function hookSafetyOutput(message) {
  return {
    hookSpecificOutput: {
      hookEventName: "UserPromptSubmit",
      additionalContext: message,
    },
  };
}

function runPromptHook(root, prompt, sessionID, context = null, { persistState = true } = {}) {
  const rootOrContext = context || root;
  const resolver = newPathResolver(rootOrContext);
  const tc = buildTimeContext(prompt, resolver, !context);
  const { response, plan } = buildWithPlan(rootOrContext, prompt);
  const additionalContext = buildWikiAdditionalContext(tc, response, plan, resolver);
  if (persistState) writeWikiPlanState(rootOrContext, sessionID, prompt, plan, context);
  if (process.env.QDM_HARNESS_DIAG === "1" && (!context || context.capabilities?.hasStableSessionId)) {
    recordDiagnostic(rootOrContext, sessionID, prompt, additionalContext, tc, response);
  }
  return {
    ok: true,
    output: {
      hookSpecificOutput: {
        hookEventName: "UserPromptSubmit",
        additionalContext,
        contextFiles: [...response.contextFiles],
      },
    },
  };
}

function buildWikiAdditionalContext(tc, response, plan, resolver) {
  let b = "# Data Harness Context\n\n";
  b += `时间解析 JSON：\`${JSON.stringify(tc)}\`\n`;
  b += `resourceRoot: \`${resolver.resourceRoot}\`\n\n`;
  b += `Harness mode: ${plan.mode}\n`;
  if (plan.selectedPlaybook) b += `selectedPlaybook: ${plan.selectedPlaybook}\n`;
  if (plan.selectedTemplate) b += `selectedTemplate: ${plan.selectedTemplate}\n`;
  if (plan.templateSelection?.status) {
    b += `templateSelection: ${plan.templateSelection.status}`;
    if (plan.templateSelection.reason) b += ` (${plan.templateSelection.reason})`;
    b += "\n";
  }
  if (plan.selectedPlaybooks?.length) {
    b += "selectedPlaybooks:\n";
    for (const playbook of plan.selectedPlaybooks) b += `- ${playbook.path}\n`;
  }
  if (plan.mode === MODE_FREE) b += `reason: ${plan.reason || ""}\n`;
  b += "\n必须先读取以下 contextFiles（可信绝对路径，不得按 workspaceRoot 解析）：\n";
  for (const ref of response.contextFiles) {
    b += `- \`${resolver.resolve(ref.path)}\``;
    const annotations = [];
    if (ref.reason) annotations.push(ref.reason);
    annotations.push(`logical: \`${ref.path}\``);
    b += ` (${annotations.join("; ")})\n`;
  }
  b += `\nInstruction: ${response.instruction}`;
  b += "\n\nConstraints:\n";
  for (const constraint of response.constraints) b += `- ${constraint}\n`;
  return b;
}

function parsePromptPayload(input) {
  try {
    const text = Buffer.isBuffer(input) ? input.toString("utf8") : String(input || "");
    const payload = JSON.parse(text);
    if (!payload || typeof payload !== "object") return null;
    return payload;
  } catch {
    return null;
  }
}

function buildTimeContext(prompt, resolver, includePrompt = true) {
  let current;
  const currentDate = process.env.QDM_HARNESS_CURRENT_DATE || "";
  if (/^\d{4}-\d{2}-\d{2}$/.test(currentDate)) current = currentDate;
  else current = formatDate(new Date());
  let timezone = process.env.QDM_HARNESS_TIMEZONE || process.env.TZ || "";
  if (!timezone) timezone = "Asia/Shanghai";
  let timePolicy = "spec/common/time-policy.md";
  try {
    const info = statSync(resolver.resolve("rules/QDM 时间口径/spec.md"));
    if (!info.isDirectory()) timePolicy = "rules/QDM 时间口径/spec.md";
  } catch {
    // keep default
  }
  const context = {
    current_date: current,
    timezone,
    time_policy: `Use ${resolver.resolve(timePolicy)} to infer --date, --week, or --month. Do not use date ranges.`,
  };
  if (includePrompt) context.prompt = prompt;
  return context;
}

function formatDate(d) {
  const y = d.getFullYear();
  const m = String(d.getMonth() + 1).padStart(2, "0");
  const day = String(d.getDate()).padStart(2, "0");
  return `${y}-${m}-${day}`;
}

function hookSessionID(payload) {
  if (payload.session_id) return payload.session_id;
  return process.env.CLAUDE_SESSION_ID || "unknown";
}

function writeWikiPlanState(root, sessionID, prompt, plan, context = null) {
  if (context && (!context.capabilities?.hasStableSessionId || !context.stateRoot)) return;
  const state = loadState(root, sessionID);
  state.mode = plan.mode;
  if (context) {
    delete state.prompt;
    state.prompt_sha256 = createHash("sha256").update(prompt).digest("hex");
  } else {
    state.prompt = prompt;
  }
  state.started_at = new Date().toISOString();
  state.playbook_candidates = plan.candidates || [];
  state.selected_playbook = "";
  state.selected_template = "";
  state.selected_playbooks = undefined;
  state.composite = undefined;
  state.reason = "";
  state.template_injected = false;
  state.reports = {};
  switch (plan.mode) {
    case MODE_SINGLE:
      state.selected_playbook = plan.selectedPlaybook;
      state.selected_template = plan.selectedTemplate;
      break;
    case MODE_MULTI:
      state.selected_playbooks = [...(plan.selectedPlaybooks || [])];
      break;
    case MODE_REPORT:
      state.selected_playbook = plan.selectedPlaybook;
      state.selected_template = plan.selectedTemplate;
      break;
    case MODE_FREE:
      state.reason = plan.reason;
      break;
    default:
      break;
  }
  saveState(root, sessionID, state);
}

function recordDiagnostic(root, sessionID, prompt, context, tc, response) {
  const event = {
    ts: new Date().toISOString(),
    session_id: sessionID,
    event: "user_prompt_context",
    context_files: response.contextFiles,
    prompt_bytes: Buffer.byteLength(prompt),
    context_bytes: Buffer.byteLength(context),
    time_context: tc,
  };
  const dir = diagnosticsDir(root);
  mkdirSync(dir, { recursive: true });
  appendFileSync(path.join(dir, `${safeSessionId(sessionID)}.jsonl`), `${JSON.stringify(event)}\n`);
}

function shouldAutoRecall(context, prompt) {
  if (String(context?.host || "").trim().toLowerCase() !== "codex") return false;
  try {
    return buildWithPlan(context, prompt).plan.mode !== MODE_FREE;
  } catch {
    return false;
  }
}

/**
 * Structured plugin hooks default to on-demand context. Codex prompts that
 * match a known Harness plan are still injected automatically; legacy
 * string-root callers retain the historical auto-context behavior.
 */
function isOnDemandContext(context, prompt, sessionID, root) {
  if (!context) return false;
  const mode = String(process.env.QDM_HARNESS_HOOK_MODE || "on-demand").trim().toLowerCase();
  if (mode === "auto-context") return false;
  if (isExplicitContextPrompt(prompt)) return false;
  if (isContinuationPrompt(prompt)) {
    try {
      const prior = loadState(context || root, sessionID);
      if (prior?.mode) return false;
    } catch {
      // If state cannot be read, fail closed and do not inject context.
    }
  }
  return true;
}

export function isExplicitContextPrompt(prompt) {
  const text = String(prompt || "").trim();
  if (!text) return false;
  if (/<skill\b[^>]*name\s*=\s*["'](?:qdm-harness|html-report)["']/i.test(text)) return true;
  if (/(?:^|[\s/])skill\s*:\s*(?:qdm-harness|html-report)\b/i.test(text)) return true;
  if (/\b(?:qdm-harness|html-report)\b/i.test(text)) return true;
  return text.includes("报告") && /生成|做|制作|输出|写|创建|来一份|周例会|经营分析|盈利情况|销售情况|分析报告/.test(text);
}

export function isContinuationPrompt(prompt) {
  return /继续|确认|已保存|保存好了|保存完|下一步|往下|推进|开始生成|开始取数|\badvance\b/i.test(String(prompt || ""));
}
